packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart

```diff
@@ -1,15 +1,338 @@
+  }) : super (
+    key: key,
+    axisDirection: axisDirection,
+    controller: controller,
+    physics: physics,
+    viewportBuilder: viewportBuilder,
+  );
+
+  final double itemExtent;
+
+  @override
+  _FixedExtentScrollableState createState() => new _FixedExtentScrollableState();
+}
+
+/// This [ScrollContext] is used by [_FixedExtentScrollPosition] to read the
+/// prescribed [itemExtent].
+class _FixedExtentScrollableState extends ScrollableState {
+  double get itemExtent {
+    // Downcast because only _FixedExtentScrollable can make _FixedExtentScrollableState.
+    final _FixedExtentScrollable actualWidget = widget;
+    return actualWidget.itemExtent;
+  }
+}
+
+/// A snapping physics that always lands directly on items instead of anywhere
+/// within the scroll extent.
+///
+/// Behaves similarly to a slot machine wheel except the ballistics simulation
+/// never overshoots and rolls back within a single item if it's to settle on
+/// that item.
+///
+/// Must be used with a scrollable that uses a [FixedExtentScrollController].
+///
+/// Defers back to the parent beyond the scroll extents.
+class FixedExtentScrollPhysics extends ScrollPhysics {
+  /// Creates a scroll physics that always lands on items.
+  const FixedExtentScrollPhysics({ ScrollPhysics parent }) : super(parent: parent);
+
+  @override
+  FixedExtentScrollPhysics applyTo(ScrollPhysics ancestor) {
+    return new FixedExtentScrollPhysics(parent: buildParent(ancestor));
+  }
+
+  @override
+  Simulation createBallisticSimulation(ScrollMetrics position, double velocity) {
+    assert(
+      position is _FixedExtentScrollPosition,
+      'FixedExtentScrollPhysics can only be used with Scrollables that uses '
+      'the FixedExtentScrollController'
+    );
+
+    final _FixedExtentScrollPosition metrics = position;
+
+    // Scenario 1:
+    // If we're out of range and not headed back in range, defer to the parent
+    // ballistics, which should put us back in range at the scrollable's boundary.
+    if ((velocity <= 0.0 && metrics.pixels <= metrics.minScrollExtent) ||
+        (velocity >= 0.0 && metrics.pixels >= metrics.maxScrollExtent)) {
+      return super.createBallisticSimulation(metrics, velocity);
+    }
+
+    // Create a test simulation to see where it would have ballistically fallen
+    // naturally without settling onto items.
+    final FrictionSimulation testFrictionSimulation =
+        // 0.135 is an arbitrary number copied from scroll_simulation.
+        new FrictionSimulation(0.135, metrics.pixels, velocity);
+
+    // Scenario 2:
+    // If it was going to end up past the scroll extent, defer back to the
+    // parent physics' ballistics again which should put us on the scrollable's
+    // boundary.
+    if (testFrictionSimulation.finalX < metrics.minScrollExtent
+        || testFrictionSimulation.finalX > metrics.maxScrollExtent) {
+      return super.createBallisticSimulation(metrics, velocity);
+    }
```

Comment:
if this gets composed with a physics class that doesn't do ballistic flinging, or does friction with a different coefficient, none of the above will work right.

Reply:
Ah crap, good catch. It would break. Done.